### PR TITLE
Support variable vhost substitution in tag_queries

### LINKED ISSUE #13: vhost variable substitution in tag_queries
Hello,

I think it would be very useful to have vhost variable substitution in tag_queries, as we have for vhost_access_query and resource_access_query.

I am not much versed in erlang, but I will try to make a PR for this. Looking at the code, it seems it can be simple.

I posted a message in rabbitmq-user about this. Here is a transcript of the scenario I described there where this would be useful:

(thread: https://groups.google.com/forum/#!topic/rabbitmq-users/VU0XAMSCeoc)

"... consider the following scenario.
- I have multiple applications on rabbitmq, each one has its own rabbitmq vhost.
- Each application has one user that can read/write/configure resources (named "sa${vhost}").
- Each application has different developers (same company, though), and I want them to manage only their own vhost (manage => readonly view of the vhost and its resources, in the production environment).
- I have a developer group per application (named "${vhost}_devs).

So i tried acls like these:

```
    {vhost_access_query, {'or', [
      {equals, {string, "${username}"}, {string, "sa${vhost}"}},
      {in_group, "cn=${vhost}_devs,ou=Users,dc=rabbittest,dc=local", "memberUid"}
    ]}},
    {resource_access_query, {equals, {string, "${username}"}, {string, "sa${vhost}"}}},
    {tag_queries, [
      {administrator, {in_group, "cn=rabbitmq_admins,ou=Users,dc=rabbittest,dc=local", "memberUid"}},
      {management, {in_group, "cn=${vhost}_devs,ou=Users,dc=rabbittest,dc=local", "memberUid"}},
    ]},
```

And then realized that it doesn't work because tag_queries does not substitute ${vhost}.

The way it is, I either :
(a) List all developer groups in the acls with 'or' (requiring a rabbitmq restart for them to manage a new vhost)
(b) Have an extra group with all developers inside it (nested groups could make this look good, but I suppose making nested groups work would be something between a pain and impossible).
(c) Set management tag to constant true, and let non-developers see an almost "empty" UI, because they don't have access to any vhost.

(obs: unfortunately, ldap basedn can't help me in this case. The developers (or even IT) in the company are spread on OUs and their common root has all company users)

Am I missing something?
Is (c) commonplace for those using authorization in the ldap plugin? It seems so much better if they could not login at all."

### PR DESCRIPTION
**Pull request by Ayanda-D**
Fixes #13.

### CODE REVIEW

**Comment by Ayanda-D** [comment]
Just a few aspects on the tests which will need updating, following implementation of #16 tests.

**Comment by michaelklishin** [comment]
Not assigning anyone as this will be reviewed by @dcorbacho, @hairyhum, and me.

**Comment by hairyhum** [comment]
Using query from tests `{management, {exists, "ou=${vhost},ou=vhosts,dc=example,dc=com"}}` authentication attempt without VHost failing with `noSuchObject` error.
Since we cannot provide VHost for AMQP connections - this will result with errors using this query.

**Comment by Ayanda-D** [comment]
@hairyhum yes indeed. I'll look into this. What's also still outstanding is a failure test to expose this scenario which you've just raised. Thank you for raising this :+1:

**Comment by hairyhum** [comment]
Looks like my previous comment is covered by https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/pull/28

**Comment by Ayanda-D** [comment]
@hairyhum okay cool, so no need for any custom handling of variable tag queries when coming from AMQP, as they'll still be eventually be deemed as "failed tag queries", as in #28. :+1:

**Comment by hairyhum** [comment]
@Ayanda-D can you review merge? There were some conflicts in test module.

**Comment by Ayanda-D** [comment]
@hairyhum thanks, can you just double check. this should now be okay.

File: test/src/rabbit_auth_backend_ldap_test.erl [hunk 1]
```diff
@@ -172,20 +172,31 @@ login() ->
...
+     {good, [5, 7], ?ALICE},
+     {good, [5, 7], ?BOB},
+     {good, [1, 2, 3, 4, 6, 7], ?PETER}].
+
+logins_direct() ->
+    [{bad,  [5], #amqp_params_direct{}},
+     {bad,  [5], #amqp_params_direct{username      = <<?ALICE_NAME>>}},
+     {bad,  [5], #amqp_params_direct{username      = <<?ALICE_NAME>>,
+                                     password      = <<"password">>}},
+     {good, [5], #amqp_params_direct{username      = <<?ALICE_NAME>>,
```
Comment (hairyhum) [inline test/src/rabbit_auth_backend_ldap_test.erl]: @Ayanda-D this test passes without https://github.com/rabbitmq/rabbitmq-server/pull/722 
Is it correct?

Comment (Ayanda-D) [inline test/src/rabbit_auth_backend_ldap_test.erl]: @hairyhum previously, it should've failed. But this is no longer the case due to #28.

Comment (hairyhum) [inline test/src/rabbit_auth_backend_ldap_test.erl]: @Ayanda-D so we need this test to check tags then.

**Pull request merged**